src/ic-cdk/src/api/stable.rs

```diff
@@ -161,6 +162,142 @@ impl io::Write for StableWriter {
     }
 }
 
+/// A writer to the stable memory which first writes data to a buffer and flushes the buffer to
+/// stable memory each time it becomes full. This reduces the number of system calls to
+/// `stable64_write` and `stable64_grow` which have relatively large overhead.
+struct BufferedStableWriter<M: StableMemory = CanisterStableMemory> {
+    /// The offset of the next write.
+    offset: u64,
+
+    /// The capacity, in pages.
+    capacity: u64,
+
+    /// The buffer to hold data waiting to be written to stable memory
+    buffer: Vec<u8>,
+
+    /// The stable memory implementation
+    memory: M,
+}
+
+impl Default for BufferedStableWriter {
+    fn default() -> Self {
+        BufferedStableWriter::new(1024 * 1024) // 1MB buffer
+    }
+}
+
+impl BufferedStableWriter {
+    pub fn new(buffer_size: usize) -> BufferedStableWriter {
+        let memory = CanisterStableMemory::default();
+
+        Self::with_memory(buffer_size, memory)
+    }
+}
+
+impl<M: StableMemory> BufferedStableWriter<M> {
+    pub fn with_memory(buffer_size: usize, memory: M) -> BufferedStableWriter<M> {
+        BufferedStableWriter {
+            offset: 0,
```

Comment:
Should `offset` be input of this function? So that the caller is aware that all bytes in the stable memory starting from `offset` are going to be overwritten?

Reply:
Done.